### PR TITLE
perf(core): binary search for findIndex

### PR DESCRIPTION
**Pull request by onx2**
This is a quick win that should help with `findIndex` performance. RE: #488

Here are a few benchmarks that I ran to get an idea of the perf diff (MS averaged over 10,000 iterations)

**Cache size: 10k | Item size: 90 | offsetToFind: 5000**
image

**Cache size: 1,000 | Item size: 45 | offsetToFind: 900**
image

**Cache size: 100 | Item size: 15 | offsetToFind: 99**
image

### CODE REVIEW

File: src/core/cache.ts [hunk 1]
```diff
@@ -217,7 +245,7 @@ export const updateCacheLength = (
     ).reduce(
       (acc, removed) =>
         acc - (removed === UNCACHED ? cache._defaultItemSize : removed),
-      0
+      0,
```
Comment (onx2) [inline src/core/cache.ts:248]:
```suggestion
      0
```

File: src/core/cache.ts [hunk 2]
```diff
@@ -95,20 +95,48 @@ export const computeTotalSize = (cache: Cache): number => {
...
+  cache: Cache,
+  offset: number,
+  initialIndex: number
+): number => {
+  let low = 0;
+  let high = cache._length - 1;
+  let mid, itemOffset;
+
+  // Adjust the search range based on the initialIndex
+  if (isValidCacheIndex(cache, initialIndex)) {
```
Comment (onx2) [inline src/core/cache.ts:115]:
```suggestion
  if (Number.isInteger(index) && index >= 0 && index < cache._length) {
```

File: src/core/cache.ts [hunk 3]
```diff
@@ -95,20 +95,48 @@ export const computeTotalSize = (cache: Cache): number => {
 /**
  * @internal
  */
-export const findIndex = (cache: Cache, offset: number, i: number): number => {
-  while (i >= 0 && i < cache._length) {
-    const itemOffset = computeOffset(cache, i);
+function isValidCacheIndex(cache: Cache, index: number) {
+  return Number.isInteger(index) && index >= 0 && index < cache._length;
+}
```
Comment (onx2) [inline src/core/cache.ts:100]:
```suggestion
```

File: src/core/cache.ts [hunk 4]
```diff
@@ -95,20 +95,45 @@ export const computeTotalSize = (cache: Cache): number => {
 /**
  * @internal
  */
```
Comment (onx2) [inline src/core/cache.ts:97]:
```suggestion
```

File: src/core/cache.ts [hunk 5]
```diff
@@ -92,23 +92,45 @@ export const computeTotalSize = (cache: Cache): number => {
   );
 };
 
+
```
Comment (onx2) [inline src/core/cache.ts:95]:
```suggestion
```

File: src/core/cache.ts [hunk 6]
```diff
@@ -96,19 +96,30 @@ export const computeTotalSize = (cache: Cache): number => {
...
+        return clamp(mid, 0, cache._length - 1);
       }
+      low = mid + 1;
     } else {
-      i--;
+      high = mid - 1;
     }
   }
-  return clamp(i, 0, cache._length - 1);
+  return clamp(low, 0, cache._length - 1);
```
Comment (onx2) [inline src/core/cache.ts:122]: Does returning with `low` make sense here? @inokawa

Comment (inokawa) [reply src/core/cache.ts:122]: Do you mean this path is unreachable? If so we should just throw error. In the past, there were some cases the index exceeded the limit, but now I'm not sure.

Comment (onx2) [reply src/core/cache.ts:122]: I think it may be possible to not find an index or have invalid input args; usually I lean toward early exit with `-1` on invalid input args and a default return of `-1` for binary search stuff. FWIW though, there isn't anything in place at the moment in the consumers of `findIndex` to handle this behavior. 

I'm happy to discuss approaches or implement your preferred approach to that issue in this PR, but ATM I think this should mimic the current behavior with just a change in how the index is searched for.

Comment (onx2) [reply src/core/cache.ts:122]: In the previous implementation, though logically it may not be possible given the constraints on input args, it could exit with `i=0` or `i=cache._length`, both of which are not a "found index" technically so it may lead to undesirable behavior.

If we were to return `-1` when no index is found we could check for this in the consumer and prevent subsequent behavior.

**Comment by onx2** [comment]
@inokawa I made some updates I thought were sensible, mind taking a look?

- jsdocs so it is easier for contributors to understand what is happening in these internal fns
- input arg validation in `findIndex`
- `-1` return when invalid input or no index found in `findIndex`
- handle invalid cases in `computeRange`

**Comment by inokawa** [comment]
Could you fix the logic so that the tests pass?

File: src/core/cache.ts [hunk 7]
```diff
@@ -94,39 +94,88 @@ export const computeTotalSize = (cache: Cache): number => {
...
-    // Clamp because prevStartIndex may exceed the limit when children decreased a lot after scrolling
-    min(prevStartIndex, cache._length - 1)
-  );
-  return [start, findIndex(cache, scrollOffset + viewportSize, start)];
+  if (cache._length <= 0) return [0, 0];
+  const lastIndex = cache._length - 1;
+  
+  // Clamp because prevStartIndex may exceed the limit when children decreased a lot after scrolling
+  const start = findIndex(cache, scrollOffset, min(prevStartIndex, lastIndex));
+  if (!isValidIndex(start, cache)) return [0, 0];
```
Comment (onx2) [inline src/core/cache.ts:171]: Maybe this should be `[0, lastIndex]`? 🤔

**Comment by onx2** [comment]
I went ahead and reverted those changes. I think there will either need to be a breaking change or more complex logic to meet the test criteria than is probably worth to change the return value of `findIndex`

**Comment by onx2** [comment]
Friendly ping on this so it doesn't get too out of date with main. @inokawa Do you have anything you'd like me to change or test before merging in?

**Comment by inokawa** [comment]
Thank you for pinging! Basically I think it's ok if PR passes all of the current tests because they cover the most of edge cases. However I'd like to check it later when I have time, to publish it without any regression.

**Comment by onx2** [comment]
> Thank you for pinging! Basically I think it's ok if PR passes all of the current tests because they cover the most of edge cases. However I'd like to check it later when I have time, to publish it without any regression.

👋🏼 Hello again! I don't mean to be a bother but would like to see this merged. Is there anything I can do to help test to get the level of confidence required for a merge?

I ran the storybook and tested a quite a few of the stories without issue but IDK if there is something specific you'd like to test. Happy to help wherever I can, I know it is very time consuming maintaining OSS!

**Review by inokawa — APPROVED** [review]
Thanks! And sorry for my late reply.
